### DISCUSSION TITLE
NextJS - App Router: What's the right way to implement a fallback page?

### DISCUSSION THREAD

**Discussion by liftgoals** [serwist/serwist#174] [Q&A]
Apologies if this is a dumb question, and I appreciate the work you do in developing this library for PWAs :)

I've been trying to make my nextJS with App Router application into a PWA with offline capabilities, and Serwist is what has brought me closer to it. But I have not been able to correctly implement the fallback page for when visiting a non-cached page while in offline usage.

I currently am successfully precaching the fallback page, and the snippet below is the closest I've been to having the redirect work, but once the page redirects it starts flickering infinitely and apparently an error is: "The service worker navigation preload request was cancelled before 'preloadResponse' settled. If you intend to use 'preloadResponse', use waitUntil() or respondWith() to wait for the promise to settle."

```
declare const self: ServiceWorkerGlobalScope;

// Ensure that `self.__SW_MANIFEST` is defined.
const SW_MANIFEST = self.__SW_MANIFEST || [];

// Add custom precache entries.
const custonPrecache = [
  { url: "/fallback", revision: "00000001" },
];

// Merge the default precache entries with the custom ones.
const precacheEntries = [ ...SW_MANIFEST, ...custonPrecache ];

const serwist = new Serwist({
  precacheEntries: precacheEntries,
  // other config
  fallbacks: {
    entries: [
      {
        url: "/fallback",
        matcher({ request }) {
          return request.mode === "navigate";
        },
      },
    ],
  },
});
```

What's even more weird is that I get the same error/behaviour when I purposely try to access my fallback page even if not offline. But obviously browsing the other pages does not give this error.

**Comment by liftgoals** [comment]
lol yeah, I guess it was a dumb question, I was using a nextJS page as my fallback page and it seems like those can't be pre-cached?

After trying different approaches I made my fallback page a plain .html file in the public directory and now it works fine. I don't even need to precache it because it's already in the public directory, which is precached by default with the self.__SW_MANIFEST.

**Comment by DuCanhGH** [comment]
@liftgoals see https://github.com/serwist/serwist/discussions/159.

Besides precaching, you may also cache the offline fallback as the service worker is being installed:

```ts
const serwist = new Serwist(options);

const urlsToPrecache = ["/~offline"] as const;

self.addEventListener("install", (event) => {
  const requestPromises = Promise.all(
    urlsToPrecache.map((entry) => {
      return serwist.handleRequest({ request: new Request(entry), event });
    }),
  );

  event.waitUntil(requestPromises);
});

serwist.addEventListeners();
```

Hope this helps!
